<a href="https://colab.research.google.com/github/sleeptiger/Abc-project/blob/main/Numpy%EC%A0%95%EB%A6%ACL_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 라이브러리란?
>- 모듈은 파이썬 확장자로 이루어져 있는 실행 가능한 하나의 파일이다.
>- 모듈안에는 한개 이상 함수, 클래스 변수가 있다.
>- 라이브러리는 `import` 로 호출한다.



In [1]:
import numpy as np # as np = numpy를 np로 사용이 가능하다.

# Numpy란?


## Numpy 
>- 다차원 배열을 쉽고 효율적으로 계산할 수 있도록 해주는 python 라이브러리
<br>
<br>
#### why Numpy
>- 데이터는 이미지, 오디오, 텍스트 등 다양하다.
>- 모든 데이터는 결국, 숫자의 배열 형태로 해석이 가능하다.
>- Numpy는 python에서 모든 데이터를 자료형으로 분석하고 계산하기 위한 라이브러리이다.

## **1. 배열이란?**
1. 같은 타입을 가지는 데이터가 나열된 집합
2. 축의 개수에 따라 차원이 달라진다.
<br>
<br>
>- axis0 = 세로/ axis1= 가로/ axis3 = 깊이

In [ ]:
silver = np.array([[1,2,3],
                  [4,5,6]])
np.shape(silver) # 세로로 2줄 가로로 3줄

(2, 3)

In [ ]:
abc = np.array([[[1,2,3],
               [4,5,6]],
               [[7,8,9],
               [10,11,12]]])
np.shape(abc)

(2, 2, 3)

## **2.배열과 슬라이싱, 정렬**
1. 배열생성
    >- 리스트를 np로 변경 
    >- np를 사용해서 생성 -> 타입설정 가능
    >- 파일기반 배열 생성

In [6]:
#리스트를 np로 변경
x = [1,2,3,4]
print(type(x))
y = np.array(x)
print(type(y))

<class 'list'>
<class 'numpy.ndarray'>


In [ ]:
#   np.arange(x,y,z)   -- x = 시작 하는 숫자
#                         Y = 끝나는 숫자(Y미만의 숫자)
#                         Z = 간격

In [8]:
np.linspace(0,10,8)   # linspace(x,y,z) - z숫자 만큼 배열 시킨다.

array([ 0.        ,  1.42857143,  2.85714286,  4.28571429,  5.71428571,
        7.14285714,  8.57142857, 10.        ])

In [ ]:
# 여러가지 NUMPY 배열 생성기 
np.zeros((3,5))
# axis0 = 3  axis1= 5

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [9]:
np.full([3,3],10)
#   axis0 =3 axis1 = 3 채워주는 숫자는 10

array([[10, 10, 10],
       [10, 10, 10],
       [10, 10, 10]])

### 슬라이싱
- 특정 구간을 불러 들인다.
- `,` 중심으로 (x,y,z) = (axis0, axis1, axis3) = (행 인덱스, 열 인덱스, N축 인덱스)
-  `:` = 전체를 뜻함 ex) eve[:,1] = 모든 행에 1번 열 인덱스 출력


In [ ]:
eve = np.array([[1,2,3],
                [4,5,6],
                [7,8,9]])

In [ ]:
 np.array(eve[:,1])

array([2, 5, 8])

In [ ]:
eve[0,2]

3

### FANCY 인덱싱
- 리스트로서 특정 행/ 열을 슬라이싱 할 수 있다.

In [ ]:
arr=np.zeros((8,8))

In [11]:
arr1 =np.arange(1,17).reshape(4,4)
arr1

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [17]:
x = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
x[1:8:2] # array[x:y:z] = x= 시작 구간/ y = 끝나는 구간/ z = 간격

array([1, 3, 5, 7])

In [22]:
arr1[:,::2]

array([[ 2,  4],
       [ 6,  8],
       [10, 12],
       [14, 16]])

In [ ]:
arr1[::2] #  행 기준 2개 겨격 출력

array([[ 1,  2,  3,  4],
       [ 9, 10, 11, 12]])

In [12]:
arr1[1::2, [0,2]] # arr2[x:y:z, xz]  = x는 행의 시작점, z는 간격 수/ xz=열 컬럼

array([[ 5,  7],
       [13, 15]])

In [ ]:
matrix = np.zeros((8,8))
matrix[::2,[1,3,5,7]] = 1
matrix[1::2,[0,2,4,6]] = 1
np.array(matrix)

array([[0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.],
       [0., 1., 0., 1., 0., 1., 0., 1.],
       [1., 0., 1., 0., 1., 0., 1., 0.]])

### boolearn 인덱싱
- true 값만 출력

In [ ]:
arr1>8 # True/False만 출력을 원하면

array([[False, False, False, False],
       [False, False, False, False],
       [ True,  True,  True,  True],
       [ True,  True,  True,  True]])

In [ ]:
arr1[arr1>8] # True/False값을 토대로 출력

array([ 9, 10, 11, 12, 13, 14, 15, 16])

## **3.유용한 Numpy  함수**

In [ ]:
# np.min() - 최솟값
# np.max() - 최댓값
# np.argmin() -최소 인덱스 값
# np.argmax() - 최대 인덱스 값
# np.sum() - 전체 합
# np.mean() - 평균
# np.where(arr2 == 7) - arr2 에서 7의 위치를 찾는다

### diag 함수 2차원 행렬

In [ ]:
a = np.arange(1,10).reshape([3,3])
np.diag(a)

array([1, 5, 9])

In [ ]:
np.diag([1,2,3,4], k=-1) # k-1 =  행을 하나 밑으로 내린다 

array([[0, 0, 0, 0, 0],
       [1, 0, 0, 0, 0],
       [0, 2, 0, 0, 0],
       [0, 0, 3, 0, 0],
       [0, 0, 0, 4, 0]])

#### 배열 차원 변경

In [ ]:
abc1 = np.arange(0,16)
abc1

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])

In [ ]:
abc2 = abc1.reshape(4,-1)
abc2

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [ ]:
abc2_2 = abc1.reshape(4,4)
abc2_2

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

## **4.정렬 함수**     

In [ ]:
arr = np.array([[11, 9, 11, 12], 
                [8, 9, 5, 7], 
                [3, 1, 2, 4],
                [14, 16, 13, 15]])

In [ ]:
np.sort(arr, axis=0) # 위에서 아래로 높은순

array([[ 3,  1,  2,  4],
       [ 8,  9,  5,  7],
       [11,  9, 11, 12],
       [14, 16, 13, 15]])

In [ ]:
np.sort(arr, axis=0)[::-1] #위에서 아래로 낮음차순

array([[14, 16, 13, 15],
       [11,  9, 11, 12],
       [ 8,  9,  5,  7],
       [ 3,  1,  2,  4]])

In [ ]:
np.sort(arr, axis=1) # 왼쪽에서 오른쪽으로 높은 차순 정력

array([[ 9, 11, 11, 12],
       [ 5,  7,  8,  9],
       [ 1,  2,  3,  4],
       [13, 14, 15, 16]])

In [ ]:
np.sort(arr, axis=1)[::-1] # 행 뒤집고 오름 차순

array([[13, 14, 15, 16],
       [ 1,  2,  3,  4],
       [ 5,  7,  8,  9],
       [ 9, 11, 11, 12]])

## **5. 행렬 연산과 성능**

### 행열(matrix)이란
- 행과 열로 이루어진 `2차원` 배열을 행렬이라고 한다.
- 대부분 정형 데이터 세트는 행렬 형태로 이루어져 있다.
- 딥러닝 분석에서 가장 자주 쓰인다.

### 행렬의 기본 연산
-행렬의 덧셈과 뺼셈
<br>
    -> 연산되는 두 행렬의 차원이 같아야 한다.

In [ ]:
a = np.arange(1,17).reshape([4,4])
b = np.arange(1,17).reshape([4,4])
np.shape(a), np.shape(b)

((4, 4), (4, 4))

-행렬의 곱셈과 나눗셈
> Element-wise 연산 : 연산되는 두 행렬의 차원이 같아야 함
<br>
> 내적 연산(Dot Product) : 연산되는 앞행렬(열)과 뒤행렬(행) 차원이 같아야합니다.



In [ ]:
a * b # ELement-wise

array([[  1,   4,   9,  16],
       [ 25,  36,  49,  64],
       [ 81, 100, 121, 144],
       [169, 196, 225, 256]])

In [ ]:
arr = np.arange(1,7).reshape([2,3])
arr2 = np.arange(7, 13).reshape([3,2])
print(np.array(arr))
print(np.array(arr2))

[[1 2 3]
 [4 5 6]]
[[ 7  8]
 [ 9 10]
 [11 12]]


In [ ]:
np.dot(arr, arr2)    # 내적연산

array([[ 58,  64],
       [139, 154]])

<br>
58= (1 * 7)+(2 * 9)+ (3*11)
<br>
64 =(1 * 8)+(2 * 10)+ (3*12)
<br>
139 = (4 * 7)+(5 * 9)+ (6*11)
<br>
154 = (4 * 8)+(5 * 10)+ (6*12)

### 행렬 집계 함수
>- 합계(`np.sum`), 최대(`np.max`), 최소(`np.min`), 평균(`np.mean`), 표준편차(`np.std`)


In [ ]:
arr = np.arange(1,17).reshape([4,4])
arr

array([[ 1,  2,  3,  4],
       [ 5,  6,  7,  8],
       [ 9, 10, 11, 12],
       [13, 14, 15, 16]])

In [ ]:
#합계 - 전체
np.sum(arr)

136

In [ ]:
#합계 -행방향(위에서 아래로)
np.sum(arr, axis=0)

array([28, 32, 36, 40])

In [ ]:
# 합계 - 열방향 (왼쪽에서 오른쪽으로)
np.sum(arr, axis=1)

array([10, 26, 42, 58])

### Broadcasting 연산
- 저차원 배열의 `연산`을 `고차원` 배열로 확장시키는 의미
- `주의사항` 행과 열의 길이는 일치 해주어야 한다.

In [ ]:
arr3d = np.array([[[0,1],[2,3],[4,5],[6,7]], [[8,9],[10,11],[12,13],[14,15]], [[16,17],[18,19],[20,21],[22,23]]])
arr2d = np.array([[0,1],[2,3],[4,5],[6,7]])
np.shape(arr3d), np.shape(arr2d)

((3, 4, 2), (4, 2))

In [ ]:
np.array(arr3d)

array([[[ 0,  1],
        [ 2,  3],
        [ 4,  5],
        [ 6,  7]],

       [[ 8,  9],
        [10, 11],
        [12, 13],
        [14, 15]],

       [[16, 17],
        [18, 19],
        [20, 21],
        [22, 23]]])

In [ ]:
np.array(arr2d)

array([[0, 1],
       [2, 3],
       [4, 5],
       [6, 7]])

In [ ]:
# 고차원으로 맞춰서 연산 
arr3d + arr2d

array([[[ 0,  2],
        [ 4,  6],
        [ 8, 10],
        [12, 14]],

       [[ 8, 10],
        [12, 14],
        [16, 18],
        [20, 22]],

       [[16, 18],
        [20, 22],
        [24, 26],
        [28, 30]]])

In [ ]:
#모든 값에 사칙연상을 진행한다.
arr2d + 10

array([[10, 11],
       [12, 13],
       [14, 15],
       [16, 17]])

In [ ]:
arr2d * 10

array([[ 0, 10],
       [20, 30],
       [40, 50],
       [60, 70]])

In [ ]:
arr2d / 10

array([[0. , 0.1],
       [0.2, 0.3],
       [0.4, 0.5],
       [0.6, 0.7]])